src/istgt_integration_test.c

```diff
@@ -0,0 +1,726 @@
+		if (zio_cmd->buf != NULL) {
+			rc = uzfs_zvol_socket_read(rargs->iofd, zio_cmd->buf, hdr->len);
+			if (rc != 0) {
+				zio_cmd_free(&zio_cmd);
+				REPLICA_ERRLOG("Socket read failed with "
+				    "error: %d\n", errno);
+				goto end;
+			}
+		}
+
+		MTX_LOCK(&rargs->io_recv_mtx);
+		TAILQ_INSERT_TAIL(&rargs->io_recv_list, zio_cmd, next);
+		pthread_cond_signal(&rargs->io_recv_cv);
+		MTX_UNLOCK(&rargs->io_recv_mtx);
+	}
+end:
+	return NULL;
+}
+
+pthread_mutexattr_t mutex_attr;
+extern void create_mock_client(spec_t *);
+uint64_t blocklen;
+uint64_t volsize;
+
+/*
+ * Initialize mutex, cv variables in spec which are required for repliation
+ */
+static int
+initialize_spec(spec_t *spec)
+{
+	int k, rc;
+	memset(spec, 0, sizeof (spec_t));
+	spec->volname = xstrdup("vol1");
+	spec->blocklen = blocklen;
+	spec->blockcnt = (volsize / spec->blocklen);
+
+	for(k = 0; k < ISTGT_MAX_NUM_LUWORKERS; k++) {
+		rc = pthread_cond_init(&spec->luworker_rcond[k], NULL);
+		if (rc != 0) {
+			REPLICA_ERRLOG("luworker %d rcond_init() failed errno:%d\n", k, errno);
+			return -1;
+		}
+
+		rc = pthread_mutex_init(&spec->luworker_rmutex[k], &mutex_attr);
+		if (rc != 0) {
+			REPLICA_ERRLOG("luworker %d mutex_init() failed errno:%d\n", k, errno);
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/*
+ * main replica thread to accept for data connections
+ * creates other worker threads for reading/writing/executing
+ */
+static void *
+mock_repl(void *args)
+{
+	rargs_t *rargs = (rargs_t *)args;
+	int file_fd, sfd, mgmtfd;
+	pthread_t mgmt_receiver, mgmt_sender, mgmt_worker;
+	pthread_t io_receiver, io_sender, io_worker1, io_worker2, io_worker3;
+	struct sockaddr saddr;
+	socklen_t slen;
+
+	snprintf(tinfo, 50, "mock%d", rargs->replica_port);
+	prctl(PR_SET_NAME, tinfo, 0, 0, 0);
+
+	rargs->file_fd = file_fd = open(rargs->file_path, O_RDWR, 0666);
+
+	//Create listener for io connections from controller and add to epoll
+	if((sfd = cstor_ops.conn_listen(rargs->replica_ip, rargs->replica_port, 32, 0)) < 0) {
+		REPLICA_ERRLOG("conn_listen() failed, errorno:%d", errno);
+		exit(EXIT_FAILURE);
+        }
+
+	//Connect to controller to start handshake and connect to epoll
+	while((rargs->mgmtfd = mgmtfd = cstor_ops.conn_connect(rargs->ctrl_ip, rargs->ctrl_port)) < 0) {
+		REPLICA_ERRLOG("conn_connect() failed errno:%d\n", errno);
+		sleep(1);
+	}
+
+	pthread_mutex_init(&rargs->mgmt_recv_mtx, NULL);
+	pthread_mutex_init(&rargs->mgmt_send_mtx, NULL);
+
+	pthread_cond_init(&rargs->mgmt_recv_cv, NULL);
+	pthread_cond_init(&rargs->mgmt_send_cv, NULL);
+
+	TAILQ_INIT(&rargs->mgmt_recv_list);
+	TAILQ_INIT(&rargs->mgmt_send_list);
+
+	pthread_mutex_init(&rargs->io_recv_mtx, NULL);
+	pthread_mutex_init(&rargs->io_send_mtx, NULL);
+
+	pthread_cond_init(&rargs->io_recv_cv, NULL);
+	pthread_cond_init(&rargs->io_send_cv, NULL);
+
+	TAILQ_INIT(&rargs->io_recv_list);
+	TAILQ_INIT(&rargs->io_send_list);
+
+	pthread_create(&mgmt_receiver, NULL, &mock_repl_mgmt_receiver, args);
```

Comment:
Precautionary, we can arrange this code such that before we create receiver thread, we should have sender/worker threads created.